zkvm/src/blockchain/block.rs

```diff
@@ -0,0 +1,121 @@
+use crate::{MerkleTree, Tx, TxID, TxLog, Verifier};
+
+#[derive(Clone, PartialEq)]
+pub struct BlockID(pub [u8; 32]);
+
+#[derive(Clone)]
+pub struct BlockHeader {
+    pub version: u64,
+    pub height: u64,
+    pub prev: BlockID,
+    pub timestamp_ms: u64,
+    pub txroot: [u8; 32],
+    pub utxoroot: [u8; 32],
+    pub nonceroot: [u8; 32],
+    pub refscount: u64,
+    pub ext: Box<[u8]>,
+}
+
+impl BlockHeader {
+    pub fn id(&self) -> BlockID {
+        let mut t = Transcript::new(b"ZkVM.blockheader");
+        t.commit_u64(b"version", self.version);
+        t.commit_u64(b"height", self.height);
+        t.commit_bytes(b"previd", &self.prev.0);
+        t.commit_u64(b"timestamp_ms", self.timestamp_ms);
+        t.commit_bytes(b"txroot", &self.txroot);
+        t.commit_bytes(b"utxoroot", &self.utxoroot);
+        t.commit_bytes(b"nonceroot", &self.nonceroot);
+        t.commit_u64(b"refscount", self.refscount);
+        t.commit_bytes(b"ext", &self.ext);
+
+        let mut result = [0u8; 32];
+        t.challenge_bytes(b"id", &mut result);
+        BlockID(result)
+    }
+
+    pub fn make_initial(timestamp_ms: u64, refscount: u64) -> BlockHeader {
+        BlockHeader {
+            version: 1,
+            height: 1,
+            prev: BlockID([0; 32]),
+            timestamp_ms: timestamp_ms,
+            txroot: [0; 32],
+            utxoroot: [0; 32],
+            nonceroot: [0; 32],
+            refscount: refscount,
+            ext: Box::new([]),
+        }
+    }
+}
+
+pub struct Block {
+    pub header: BlockHeader,
+    pub txs: Vec<Tx>,
+}
+
+impl Block {
+    pub fn validate(&self, prev: &BlockHeader) -> Result<Vec<TxLog>, BCError> {
+        if self.header.version < prev.version {
+            return Err(BCError::VersionReversion);
+        }
+        if self.header.version == 1 && self.header.ext.len() != 0 {
+            return Err(BCError::IllegalExtension);
+        }
+        if self.header.height != prev.height + 1 {
+            return Err(BCError::BadHeight);
+        }
+        if self.header.prev != prev.id() {
+            return Err(BCError::MismatchedPrev);
+        }
+        if self.header.timestamp_ms <= prev.timestamp_ms {
+            return Err(BCError::BadBlockTimestamp);
+        }
+        if self.header.refscount > prev.refscount + 1 {
+            return Err(BCError::BadRefscount);
+        }
+
+        let mut txlogs: Vec<TxLog> = Vec::new();
+        let mut txids: Vec<TxID> = Vec::new();
+
+        for tx in self.txs.iter() {
+            if tx.header.mintime_ms > self.header.timestamp_ms
+                || self.header.timestamp_ms > tx.header.maxtime_ms
+            {
+                return Err(BCError::BadTxTimestamp);
+            }
+            if self.header.version == 1 && tx.header.version != 1 {
+                return Err(BCError::BadTxVersion);
+            }
+
+            // TODO(bobg): The API currently requires anticipating how many multipliers (generators) are needed,
```

Comment:
Based on conversation in the Dalek slack with Henry and Vicki, the best path forward is:

1. On the prover, measure how many mults you need from a proof, and insert that into Tx object. We can also have a “measurement” impl of ConstraintSystem trait that collects that w/o fully computing the r1cs proof, but that’s more like an optimization.
2. Verifier would have the required capacity in the Tx object immediately, so it can ensure necessary capacity of the generators or fail early if it’s beyond whatever limits.

Reply:
I'm not 100% sure but it sounds like this is not actionable feedback about this PR, so I'm resolving it.

Reply:
Yeah, sorry, this is not actionable. Need to file an issue for this.